models/resnet/resnet_impl.hpp

```diff
@@ -0,0 +1,185 @@
+/**
+ * @file resnet_impl.hpp
+ * @author Aakash Kaushik
+ *
+ * Implementation of ResNet using mlpack.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MODELS_MODELS_RESNET_RESNET_IMPL_HPP
+#define MODELS_MODELS_RESNET_RESNET_IMPL_HPP
+
+#include "resnet.hpp"
+
+namespace mlpack {
+namespace models {
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet() :
+    inputChannel(0),
+    inputWidth(0),
+    inputHeight(0),
+    numClasses(0)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet(
+    const size_t inputChannel,
+    const size_t inputWidth,
+    const size_t inputHeight,
+    const bool includeTop,
+    const bool preTrained,
+    const size_t numClasses) :
+    ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>(
+        std::tuple<size_t, size_t, size_t>(
+        inputChannel,
+        inputWidth,
+        inputHeight),
+        includeTop,
+        preTrained,
+        numClasses)
+{
+  // Nothing to do here.
+}
+
+template<typename OutputLayerType, typename InitializationRuleType,
+    size_t ResNetVersion>
+ResNet<OutputLayerType, InitializationRuleType, ResNetVersion>::ResNet(
+    std::tuple<size_t, size_t, size_t> inputShape,
+    const bool includeTop,
+    const bool preTrained,
+    const size_t numClasses) :
+    inputChannel(std::get<0>(inputShape)),
+    inputWidth(std::get<1>(inputShape)),
+    inputHeight(std::get<2>(inputShape)),
+    numClasses(numClasses)
+{
+  if (preTrained)
+  {
+    std::string home = getenv("HOME");
+    preTrainedPath = home + "/.cache/mlpack/models/weights/resnet/resnet" +
+        std::to_string(ResNetVersion) + ".bin";
+    if (Utils::PathExists(preTrainedPath, true) == false)
+    {
+      std::cout << "Downloading resnet" + std::to_string(ResNetVersion) +
```

Comment:
Might be useful to switch to `mlpack::Log` which can be enabled/disabled by the user.

Reply:
It was intentional because, this only occurs when the user doesn't has the model and if that is not there curl starts a download bar so it prompts the user that this file has been downloaded to this place.

Reply:
Sure, does that apply to all the` std::cout` you used int he code?

Reply:
I believe for the new code this is the only place i have used `std::cout`.

Reply:
Right, I thought I have seen it in other places as well, but you are right this is the only place you used `cout`.

Reply:
You might have seen some previous code, a lot of the things wete already shifted to `Log`.